File: cpp/easy/ValidPalindrome.cpp
#include <string>

using namespace std;
bool isPalindrome(string s, int left, int right, bool chance) {
  if (left == right) {
    return true;
  }
  while (left < right) {
    if (s[left] != s[right]) {
      if (chance) {
        return isPalindrome(s, left, right - 1, false) ||
               isPalindrome(s, left + 1, right, false);
      }
      return false;
    }
    right--;
    left++;
  }
  return true;
}
bool validPalindrome(string s) {
  return isPalindrome(s, 0, s.size() - 1, true);
}

File: go/easy/validPalindrome.go
package main

func checkPalindrome(s string, left int, right int, canRemove bool) bool {
	for left < right {
		if s[right] == s[left] {
			right--
			left++
			continue
		} else if canRemove {
			return checkPalindrome(s, left+1, right, false) || checkPalindrome(s, left, right-1, false)
		} else {
			return false
		}
	}
	return true
}

func validPalindrome(s string) bool {
	return checkPalindrome(s, 0, len(s)-1, true)
}
